python/vision_explanation_methods/error_labeling/error_labeling.py

```diff
@@ -0,0 +1,229 @@
+                    continue
+                if (self._iou_threshold <= iou_score):
+                    # the detection and ground truth bb's are overlapping
+                    if detect[0] != gt[0]:
+                        # the bb's line up, but labels do not
+                        self._match_matrix[gt_index][detect_index] = (
+                            ErrorLabelType.CLASS_NAME)
+                        continue
+                    elif (ErrorLabelType.MATCH in
+                          self._match_matrix[gt_index]):
+                        # class name and bb correct, but there is already a
+                        # match with a higher confidence score (this is why
+                        # it was imporant to sort by descending confidence
+                        # scores as the first step)
+                        self._match_matrix[gt_index][detect_index] = (
+                            ErrorLabelType.DUPLICATE_DETECTION)
+                        continue
+                    else:
+                        # this means bbs overlap, class names = (1st time)
+                        self._match_matrix[gt_index][detect_index] = (
+                            ErrorLabelType.MATCH)
+                        continue
+                else:
+                    if detect[0] != gt[0]:
+                        # the bb's don't line up, and labels do not
+                        self._match_matrix[gt_index][detect_index] = (
+                            ErrorLabelType.CLASS_LOCALIZATION)
+                        continue
+                    else:
+                        # the bb's don't line up, but the labels are correct
+                        self._match_matrix[gt_index][detect_index] = (
+                            ErrorLabelType.LOCALIZATION)
+                        continue
+
+            # resort the columns (so no longer ordered by descending conf
+            # scores)
+            self._match_matrix[gt_index] = [self._match_matrix[gt_index][i]
+                                            for i in original_indices]
+
+    def compute_error_list(self):
+        """Determine a complete list of errors encountered during prediction.
+
+        Note that it is possible to have more errors than actual objects
+        in an image (because we account for missing detections and
+        duplicate detections).
+        """
+        self.compute()
+        error_arr = self._remove_matches(deepcopy(self._match_matrix))
+        dup_count = sum(1 for row in self._match_matrix for element in row
+                        if element == ErrorLabelType.DUPLICATE_DETECTION)
+        error_list = [ErrorLabelType.DUPLICATE_DETECTION
+                      for i in range(0, dup_count)]
+
+        if error_arr == []:
+            return error_list
+
+        diff = len(error_arr) - len(error_arr[0])
+        if diff > 0:
+            error_list += [ErrorLabelType.MISSING for i in range(0, diff)]
```

Comment:
Well, tbh a better approach would be to append these strings to `error_list` instead of adding 2 python lists. `append` is amortized O(1) but adding 2 dynamic arrays is a linear operation with `O(A+B)` I believe as the compiler needs to fine a new place in memory to hold all those contiguous elements.

So better recommendation would be

```suggestion
            for _ in range(diff):
                error_list.append(ErrorLabelType.MISSING)
```

Reply:
Interesting! Didn't know that!